### ISSUE TITLE
NullReferenceException when calling a method that returns a value type and has been configured with DoesNothing

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#806]
This code throws a `NullReferenceException` on the call to `Baz()`:

```c#
void Main()
{
    var foo = A.Fake<IFoo>();
    A.CallTo(foo).DoesNothing();
    foo.Bar();
    foo.Baz();
}

public interface IFoo
{
    string Bar();
    int Baz();
}
```

The stack trace isn't very helpful:
```
   at Castle.Proxies.ObjectProxy_8.Baz()
   at UserQuery.Main()
   ...
```
But the fact that `Baz` throws and `Bar` doesn't is a good hint: I think the generated code tries to cast the `IInvocation.ReturnValue` to `int`, which fails because it's null.

In a way, that's a bug in Castle.Core (it should throw a more specific exception). Maybe we could do something about it, but I'm not sure what we *should* do. I discovered this bug by accident; I didn't really want to use `DoesNothing()`, what I wanted was to leave the default behavior, but I needed to specify a behavior in order to use `.Once()`. Currently, if you just want to add a callback with `.Invokes()`, there's no way to _not_ affect the default behavior, since you have to specify a behavior. Maybe we need something like `.UseDefaultBehavior()`...

**Comment by blairconrad** [comment]
I'm surprised that's never come up yet, given the simple reproduction.

As I think about this, I'm wondering why we'd ever allow `DoesNothing` on a non-void method. I understand the need to sometimes override strictness or a previously-specified throw, but "doing nothing" isn't really an option.

Instead of `UseDefaultBehavior`, what about `ReturnsDefaultValue`? And `DoesNothing` can remain for the void methods?

I know, a breaking change. But since we're talking about it, I thought I might as well consider the API.

**Comment by blairconrad** [comment]
`ReturnsDefaultValue` might be a little confusing as well, since people could expect that it would be `default(T)`. Perhaps a better name would be in order.
Or maybe the stance is "you wanted to customize this method, so you're responsible for telling me what the return value should be".

**Comment by thomaslevesque** [comment]
> As I think about this, I'm wondering why we'd ever allow `DoesNothing` on a non-void method. I understand the need to sometimes override strictness or a previously-specified throw, but "doing nothing" isn't really an option. We have to at least return something.

In this case, we don't know it's non-void... it could be anything.

> Instead of `UseDefaultBehavior`, what about `ReturnsDefaultValue`? And `DoesNothing` can remain for the void methods?

`ReturnsDefaultValue` would only make sense for non-void methods, that's why I picked a more neutral name. In fact, `KeepCurrentBehavior` would probably be better than `UseDefaultBehavior`; it would use the previously defined behavior, whatever it is. This way, it would still pick the behavior defined in a fake options builder.

**Comment by adamralph** [comment]
Defaulting to P2.

**Comment by blairconrad** [comment]
As I continue to think about this, I find that my feelings edge more toward disallowing use of `DoesNothing` unless we know the method is void (which we could accomplish by adding the `WithVoidReturnType` to mirror `WithNonVoidReturnType`, as I think @thomaslevesque suggested elsewhere.

Then, for a method that has a return value, if you want to override some previously configured behaviour, you have to pick a new return value (or throw an exception, I suppose). I don't think it's reasonable for the framework to examine previously-defined behaviours and decide which one the user might want.

If we really want an option for the user to specify "do what you'd do if I'd never configured the method at all", what about `ReturnsADummy`? Since that's what it would do.
Of course, without this option the user could just `Returns(A.Dummy<T>())`, if they didn't have a value in mind.

**Comment by blairconrad** [comment]
I continue to continue to think about this. :grin:

Restricting `DoesNothing` to void methods is certainly an option, but I think there's another path to take (which could complement the restriction, or make us feel like we don't need the restriction).

Currently, an unconfigured method with a return value will return a `Dummy` of the appropriate type when called.

However, as soon as we start building a rule (for example, when we add an `Invokes` behaviour to the method), we create a rule that will return the "default value" for the type, as we can see in [BuildableCallRule's constructor](https://github.com/FakeItEasy/FakeItEasy/blob/40e6aefec60d17cbebe6c39c34e1089adb0881b9/src/FakeItEasy/Configuration/BuildableCallRule.cs#L23). `GetDefaultValue` [returns null for a reference type or an instance for a value type](https://github.com/FakeItEasy/FakeItEasy/blob/46becf4a6cc8d72ddefc8ca024167d29979efdde/src/FakeItEasy/TypeExtensions.cs#L33).

And then if we decide to have the method `DoesNothing`, the rule is modified so that it will [not set the return value at all](https://github.com/FakeItEasy/FakeItEasy/blob/8e950d803b0a7efc45b17cd882cd0ee7b88dda05/src/FakeItEasy/Configuration/RuleBuilder.cs#L66) (and this is the cause of the bug that @thomaslevesque found).

I think that all three situations should use the "unconfigured behaviour" and return a `Dummy` (or nothing for a void of course).

I think that clears things up. No error in the case that @thomaslevesque identified, and we have the behaviour that an unconfigured method (or a configured method that doesn't specify some other return value) returns a Dummy.

**Comment by blairconrad** [comment]
I've a nearly PR-ready implementation of the changes I suggested above, just in case people like them.
And of course, they are **breaking**.

**Comment by adamralph** [comment]
Is there no way to solve this without a breaking change? I.e. should we be labelling this issue as `breaking`?

**Comment by blairconrad** [comment]
Good point. I believe we could solve the exact problem described in the title without a breaking change by having `DoesNothing` have the same behaviour as `Invokes` with respsect to return values, that is, it would return `null` for reference types and an `Activator.CreateInstance`-created instance for value types. Essentially `DoesNothing` would mirror `Invokes(() => {})` (which is doing nothing!), but we wouldn't have to express it that way.
I'd be down for putting that in a 2.x release, although I'd like to at least consider the larger change where every unspecified return value is a Dummy, as I do enjoy consistent behaviour across the API. But that (breaking) change could certainly wait.

**Comment by blairconrad** [comment]
How's about I submit a PR just to fix the `NullReferenceException` and split off an issue for the "inconsistent return values" problem I describe above, which can potentially be fixed in a future release.

**Comment by adamralph** [comment]
:thumbsup: I'm always happy to split things up in that way.

**Comment by adamralph** [comment]
Will be released in 2.3.0.

**Comment by blairconrad** [comment]
This issue has been released in [FakeItEasy 2.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.3.0).
